From 7f0400b98eaf3a7c69a2aa835f2ed2994eecf94b Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Sat, 9 Apr 2011 15:16:33 -0600
Subject: [PATCH 53/91] - readdir req should have position where readdir should
 start at. handlers should use it for positioning in namespace table.

---
 cob/cob.c            | 7 ++++---
 cob/cob.h            | 2 ++
 mdservice/md_fops.ff | 3 ++-
 mdstore/mdstore.c    | 5 +++--
 4 files changed, 11 insertions(+), 6 deletions(-)

diff --git a/cob/cob.c b/cob/cob.c
index bcaf9b9..382c0ea 100644
--- a/cob/cob.c
+++ b/cob/cob.c
@@ -525,14 +525,15 @@ int c2_cob_locate(struct c2_cob_domain *dom, struct c2_cob_oikey *oikey,
 	return rc;
 }
 
-int c2_cob_readdir_init(struct c2_cob *cob, struct c2_db_tx *tx)
+int c2_cob_readdir_init(struct c2_cob *cob, const char *name, 
+                        int namelen, struct c2_db_tx *tx)
 {
         int rc;
 
         /*
-         * Prepare "." entry key.
+         * Prepare entry key using passed started pos.
          */        
-        c2_cob_make_nskey(&cob->co_rdkey, &cob->co_fid, ".", 1);
+        c2_cob_make_nskey(&cob->co_rdkey, &cob->co_fid, name, namelen);
 
         /*
          * Init readdir cursor.
diff --git a/cob/cob.h b/cob/cob.h
index e73da71..1e1686d 100644
--- a/cob/cob.h
+++ b/cob/cob.h
@@ -359,6 +359,8 @@ int c2_cob_update_name(struct c2_cob        *cob,
    Init readdir iterator.
 */
 int c2_cob_readdir_init(struct c2_cob *cob, 
+                        const char *name, 
+                        int namelen, 
                         struct c2_db_tx *tx);
 
 /**
diff --git a/mdservice/md_fops.ff b/mdservice/md_fops.ff
index b5ac480..59c5d61 100644
--- a/mdservice/md_fops.ff
+++ b/mdservice/md_fops.ff
@@ -92,7 +92,8 @@ DEF(c2_fop_getattr_rep, RECORD,
     _(g_body, c2_fop_cob));
 
 DEF(c2_fop_readdir, RECORD,
-    _(r_body, c2_fop_cob));
+    _(r_body, c2_fop_cob),
+    _(r_pos, c2_fop_name));
 
 DEF(c2_fop_readdir_dirent, RECORD,
     _(d_body, c2_fop_cob),
diff --git a/mdstore/mdstore.c b/mdstore/mdstore.c
index 0d0c2af..8dbbd32 100644
--- a/mdstore/mdstore.c
+++ b/mdstore/mdstore.c
@@ -661,7 +661,9 @@ int c2_md_store_readdir(struct c2_md_store *md,
         
         C2_ASSERT(cob != NULL);        
 
-        rc = c2_cob_readdir_init(cob, &ctx->fc_tx->tx_dbtx);
+        rc = c2_cob_readdir_init(cob, req->r_pos.n_name, 
+                                 req->r_pos.n_count, 
+                                 &ctx->fc_tx->tx_dbtx);
         if (rc)
                 goto out;
 
@@ -672,7 +674,6 @@ int c2_md_store_readdir(struct c2_md_store *md,
                 if (rc == 0) {
                         ent = &vec->v_dirent[vec->v_count++];
                         ent->d_name.n_count = nskey->cnk_name.b_len;
-
                         memcpy(ent->d_name.n_name, 
                                c2_bitstring_buf_get(&nskey->cnk_name),
                                nskey->cnk_name.b_len);
-- 
1.8.3.2

